### ISSUE TITLE
Localisation

### ISSUE THREAD

**Issue by p0358** [concierge/Concierge#174]
What about some functionality to add localisations and translations? For Kassy and modules, with ability to set language in config (missing strings would be taken from English and there could be placeholder string specified in code in case it's missing in all lang files) There could be lang folder for translations and they could also be put into modules' folders for custom modules. That could be good feature, as Kassy's getting bigger. I could personally translate Kassy to Polish.
I don't know how I can mark this issue as feature or enhancement, sorry, I'm pretty new to Git.

**Comment by mrkno** [comment]
Could be done for most internal things, I don't know about modules - to fully localise them would require an entire dictionary plus semantic parser for each language which is way beyond the scope of this project.

It might be possible to provide a localisation service - you provide the translations with the module (optionally) and it will select the right one for you based on user preference.

**Comment by mrkno** [comment]
While you are commenting on things, @phillipadsmith you had mentioned that it would be nice to have this (and customize internal strings). Anything you want to add before I consider adding it to the ever growing list of changes in https://github.com/concierge/Kassy/pull/175?

**Comment by phillipadsmith** [comment]
> While you are commenting on things, @phillipadsmith you had mentioned that it would be nice to have this (and customize internal strings). Anything you want to add...

Nothing much to add beyond what I was proposing in Gittr, i.e. the ability to override the built-in strings (responses to core module commands) and to provide translations for those. 

At the time, I was investigating http://i18next.com/ as the start of a solution to the problem I was working on... it's still a work in progress. :-/

**Comment by mrkno** [comment]
Supported as of https://github.com/concierge/Kassy/pull/175/commits/98524182b7c5cce0dde063522f64ec393aa1d558 in PR #175

**Comment by mrkno** [comment]
@p0358 you said you would be able to translate into polish? If you are still interested, there is a file that needs to be translated here: https://github.com/concierge/Kassy/blob/feature/general-feature-enhancements/core/translations/i18n/en.json

**Comment by phillipadsmith** [comment]
@mrkno That is a beautiful thing. :-)  So can non-core modules hook into this?

**Comment by mrkno** [comment]
@phillipadsmith yes. They can create an `i18n` directory with `.json` files similar to the one above within the module itself and it will automatically route any translations to it.

E.g. in a module called `foo`: `/modules/foo/i18n/en.json` would be the english translations.

**Comment by p0358** [comment]
`{
  "${0} failed ${1} caused it": "${0} wywalił się. ${1} jest teraz pokryty chorobą.",
  "${0} Shutdown": "${0} został(a) wyłączony(/a).",
  "[DONE]": "[GOTOWE]",
  "[FAIL]": "[BŁĄD]",
  "A failure occured while listing \"${0}\". It doesn't appear to be a module.": "Niepowodzenie podczas listowania \"${0}\". To nie wygląda na moduł.",
  "A module must provide a match function or command.": "Moduł musi dostarczyć funkcję dopasowania (match) lub komendę.",
  "A module must provide basic help.": "Moduł musi dostarczyć podstawową pomoc.",
  "An error occured while saving the configuration file.": "Wystąpił błąd podczas zapisywania pliku konfiguracyjnego.",
  "BrokenModule ${0}": "Moduł ${0} wygląda na uszkodzony. Usuń go lub napraw.",
  "Cannot change integrations when they are already set.": "Nie można zmienić integracji, gdy są już ustawione.",
  "Cannot get integrations if they have not already been set.": "Nie można pobrać integracji, jeżeli te nie zostały jeszcze ustawione.",
  "Cannot stop integrations if they haven't been started.": "Nie można zatrzymać integracji, jeżeli te nie zostały wystartowane.",
  "Configuration data for module ${0} stored in deprecated location.": "\nDane konfiguracyjne dla modułu \"${0}\" przechowywane w przestarzałej lokalizacji.\nKonfiguracja zostanie przeniesiona do pliku konfiguracyjnego specyficznego dla modułu.",
  "Could not load module '${0}'. Does it have a syntax error?": "Nie można było załadować modułu '${0}'. Czy ma on błąd w składni?",
  "Does something. The unhelpful author didn't specify what.": "Coś robi. Niepomocny autor nie sprecyzował co.",
  "Failed to correctly stop output integration '${0}'.": "Niepowodzenie przy próbie poprawnego zatrzymania wyjściowej (output) integracji '${0}'.",
  "Failed to start output integration '${0}'.": "Niepowodzenie przy próbie uruchomienia wyjściowej (output) integracji '${0}'.",
  "HelloWorld": "Witaj świecie!",
  "I also have something to send you but cant seem to do so...": "Też mam ci coś do wysłania, ale nie wygląda na to, że mogę to zrobić...",
  "I have a file to send you but cant seem to do so...": "Mam plik do wysłania, ale nie wygląda na to, że mogę go przesłać...",
  "I have something to send you but cant seem to do so...": "Mam ci coś do wysłania, ale nie wygląda na to, że mogę to zrobić...",
  "If I could set the title of this chat I would set it to \"${0}\"": "Jeśli mógłbym ustawić tytuł tego chatu, ustawiłbym go na \"${0}\"",
  "If this is your file please ensure that it is syntactically correct.": "Jeśli to jest twój plik, proszę upewnij się, że jest składniowo poprawny.",
  "Installation complete.": "Instalacja ukończona.",
  "Installing \"${0}\" from npm.": "Instalowanie \"${0}\" z npm.",
  "Integration must be .js file or directory.": "Integracja musi być plikiem .js lub katalogiem.",
  "Integrations must be set before starting": "Integracje muszą być ustawione przed startem",
  "Invalid integration installed (syntax error?).": "Nieprawidłowa integracja zainstalowana (błąd składniowy?).",
  "Invalid translation context provided.": "Nieprawidłowy kontekst został dostarczony do tłumaczenia.",
  "Loading integration '${0}'...\t": "Ładowanie integracji '${0}'...\t",
  "Loading module '${0}'... ${1}": "Ładowanie modułu '${0}'... ${1}",
  "Loading Succeeded": "Ładowanie zakończono pomyślnie",
  "Loading the output integration file '${0}' failed.\n\n": "Ładowanie pliku integracji wyjściowej '${0}' nie powiodło się.\n\n",
  "LoadingModules": "Ładowanie modułów...",
  "LoadingSystemConfig": "Ładowanie konfiguracji systemowej...",
  "Module \"${0}\" could not be loaded.": "Moduł \"${0}\" nie mógł zostać załadowany.",
  "No or invalid configuration file found at \"${0}\".": "Brak lub nieprawidłowy plik konfiguracyjny znaleziony w \"${0}\".",
  "Not enough information provided to describe integration.": "Niewystarczająca ilość informacji dostarczona, aby opisać integrację.",
  "ShutdownError": "Nie można wyłączyć platformy, gdy nie jest wystartowana.",
  "Skipping \"${0}\". It isn't a Hubot module.": "Pomijam \"${0}\". To nie jest moduł Hubot.",
  "Skipping \"${0}\". It isn't a Kassy module.": "Pomijam \"${0}\". To nie jest moduł Kassy.",
  "StartError": "Nie można wystartować, kiedy już wystartowano.",
  "StartingIntegrations": "Uruchamianie integracji...",
  "StartingSystem": "Uruchamianie systemu...",
  "Stopping integration '${0}'...\t": "Zatrzymywanie integracji '${0}'...\t",
  "SystemStarted": "System został uruchomiony.",
  "Unloading module \"${0}\" failed.": "Wyładowywanie modułu \"${0}\" nie powiodło się.",
  "Unloading module \"${0}\".": "Wyładowywanie modułu \"${0}\".",
  "What kind of shit platform is this that doesn't even support sending messages?": "Co to za gówniana platforma, która nawet nie wspiera wysyłania wiadomości?",
  "Working on it...": "Pracuję nad tym..."
}`

There it is :)

**Comment by mrkno** [comment]
![Polish Translation](https://cloud.githubusercontent.com/assets/6830312/16602698/52cd5f68-436a-11e6-9634-956bf90c97bc.png)

Appears to be working as intended.

**Comment by ParasDPain** [comment]
Do we have an immediate wish list of languages to provide? I'm proficient in 3 others besides english (japanese, hindi and punjabi)

**Comment by mrkno** [comment]
Any that anyone wants to spend time translating. I don't really mind.

**Comment by mrkno** [comment]
@p0358 I added more translations (sorry about that) starting at line 57 of the file:
https://github.com/concierge/Kassy/blob/587f401ec2ffd4a9ea988f86d7edcc04cd8d8a74/core/translations/i18n/en.json#L57

**Comment by p0358** [comment]
@mrkno Fine, but before I translate, why Polish chars aren't working? They don't even after setting chcp 1250... If you're testing this, you can use "Zażółć gęślą jaźń." sentence, it contains all of them. ;)

**Comment by mrkno** [comment]
Hmmm. I would have said a unicode font issue, but that doesn't seem to be the case. I'll look into it.

**Comment by p0358** [comment]
Ha! Works after you simply save the file as UTF-8 (maybe not so simply, because after converting in Notepad++ you need to rewrite all chars, because they get screwed, but you can just convert it and then repaste from Pastebin I think)
![image](https://cloud.githubusercontent.com/assets/5182588/16684005/e166412e-4502-11e6-942f-12738eae2a66.png)
First sentence there was ANSI-coded from original file downloaded from Git, second with correct char was rewritten after converting file to UTF-8.

**Comment by mrkno** [comment]
Ah, that would do it. I was thinking about the program and where exactly the logic would be wrong but it turns out to be human error :)

**Comment by mrkno** [comment]
PR now merged so internationalisation (the intention of this issue) is now possible.

**Comment by mrkno** [comment]
@p0358 UTF-8 problem fixed in https://github.com/concierge/Concierge/pull/193/commits/2d46f1f7f5908fd8cc9d942d23d0d1f07f543f69

`kassy.json` help strings can be translated now as well.
